### PR TITLE
Fix expectation computations in TraceEnum_ELBO by using adjoints to calculate marginals

### PR DESCRIPTION
**Pull request by ordabayevy**
Addresses https://github.com/pyro-ppl/pyro/issues/2724.
<s>Blocked by https://github.com/pyro-ppl/funsor/pull/413.</s>

This implementation avoids exponentiation of individual terms as suggested by @eb8680  [here](https://github.com/pyro-ppl/pyro/issues/2724#issuecomment-751552777).

### CODE REVIEW

**Comment by ordabayevy** [comment]
@eb8680 can you review this PR?

Notes:
- This PR fixes #2724 
- Marginals are calculated using `adjoint` module
- To get the required list of marginals dummy targets are created (similar to what you did in this [PR](https://github.com/pyro-ppl/pyro/pull/2726/commits/972bdc44ef26bcfe08807ddda2a7c36e2acab3cb))
- In order to work this PR also requires updated adjoint patterns https://github.com/pyro-ppl/funsor/pull/444

File: pyro/contrib/funsor/infer/traceenum_elbo.py
Comment (eb8680) [inline pyro/contrib/funsor/infer/traceenum_elbo.py:176]: Hmm... I think you can divide by the plate-local normalizing constants here rather than the global `logzq`
```py
logzq_local = marginals[target].reduce(funsor.ops.logaddexp, cost.input_vars - plate_vars)
log_prob = marginals[target] - logzq
```
which should lead to lower gradient variance.  I believe this is what the main implementation of `TraceEnum_ELBO` does. More generally, this would benefit from eager normalization within the backward pass of `adjoint`.

File: pyro/contrib/funsor/infer/traceenum_elbo.py
Comment (eb8680) [inline pyro/contrib/funsor/infer/traceenum_elbo.py:159]: Use `cost` as a prototype here

File: tests/contrib/funsor/test_enum_funsor.py
Comment (eb8680) [inline tests/contrib/funsor/test_enum_funsor.py:1396]: Can you add a couple more tests that work along similar lines (comparing ELBO values in models with parallel and sequential plates)? e.g. some of the following:
```
a -> ( (cij) )

( ai -> ( bij ) )

a -> ( bi -> ( cij ) )

a -> ( bi )
|
V
( cj )
```
You might be able to copy some of these from existing tests.

**Review by eb8680 — COMMENTED** [review]
Thanks for updating. Can you explain which specific changes to `adjoint` are necessary for this to work?

**Comment by ordabayevy** [comment]
> Can you explain which specific changes to `adjoint` are necessary for this to work?

The minimal required changes are updating `adjoint_binary` and `adjoint_reduce` patterns as in pyro-ppl/funsor#444 and removing the lines

```py
            if not isinstance(v, Variable):
                target_adjs[v] = bin_op(target_adjs[v], v)
```

File: pyro/contrib/funsor/infer/traceenum_elbo.py
Comment (ordabayevy) [reply pyro/contrib/funsor/infer/traceenum_elbo.py]: All three of these work (same loss and grad):

```py
log_prob = marginals[target] 
log_prob = marginals[target] - logzq
log_prob = marginals[target] - logzq_local
```

even when compared against the main `TraceEnum_ELBO`:

```py
from pyro.infer import TraceEnum_ELBO
...
    pyro_elbo = TraceEnum_ELBO(max_plate_nesting=2)
    pyro_loss = pyro_elbo.differentiable_loss(model_plate, guide_plate)
    _check_loss_and_grads(actual_loss, pyro_loss)
```

I'm still trying to understand why. Is it because in these examples guide is always normalized?

File: pyro/contrib/funsor/infer/traceenum_elbo.py
Comment (ordabayevy) [reply pyro/contrib/funsor/infer/traceenum_elbo.py]: `cost` can be a lazy expression in which case i'm not sure how to retrieve its prototype.

File: pyro/contrib/funsor/infer/traceenum_elbo.py
Comment (eb8680) [reply pyro/contrib/funsor/infer/traceenum_elbo.py]: Because guides are always normalized and all of the integration in these tests is exact (i.e. there is no sampling).

**Review by eb8680 — COMMENTED** [review]
LGTM after you fix merge conflicts and test against https://github.com/pyro-ppl/funsor/pull/445 . I think we'll have to wait until https://github.com/pyro-ppl/funsor/pull/445 lands in Funsor to merge this PR, though.

**Comment by ordabayevy** [comment]
Thanks for reviewing @eb8680 !

**Comment by eb8680** [comment]
Now that https://github.com/pyro-ppl/funsor/pull/445 has landed can you update the `funsor[torch]` dependency to the latest `master`?  Should be good to merge if tests pass after that, although performance may be disappointing due to regressions in https://github.com/pyro-ppl/funsor/pull/445.

File: setup.py
Comment (eb8680) [inline setup.py:123]: Ah, sorry, to clarify, can you pin to the commit hash of the most recent commit on `master` (`5d737584ad055cdff8d56f38e0c3dd10d0e7e80e`), as in the commented-out line above? Pinning to a commit hash rather than a tag/branch keeps Funsor changes from causing test failures in other, unrelated Pyro PRs.

**Review by eb8680 — CHANGES_REQUESTED** [review]

**Comment by ordabayevy** [comment]
There is an unrelated to this PR `HTTPError: HTTP Error 403: Forbidden` from `cvae/main.py`.

**Review by fritzo — DISMISSED** [review]
I'd like to postpone merging this until after Pyro's 1.6 release, so as to avoid the need to release funsor

**Review by eb8680 — APPROVED** [review]

**Pull request merged**
